plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -103,7 +104,27 @@ makeLenses ''TypeCheckEnv
 
 -- | The type checking monad that the type checker runs in.
 -- In contains a 'TypeCheckEnv' and allows to throw 'TypeError's.
-type TypeCheckM uni fun cfg err = ReaderT (TypeCheckEnv uni fun cfg) (ExceptT err Quote)
+type TypeCheckT uni fun cfg m = ReaderT (TypeCheckEnv uni fun cfg) m
+
+-- | The constraints that are required for kind checking.
+type MonadKindCheck err term uni fun ann m =
+    ( MonadError err m                  -- Kind/type checking can fail
+    , AsTypeError err term uni fun ann  -- with a 'TypeError'.
+    , ToKind uni                        -- For getting the kind of a built-in type.
+    )
+
+-- | The general constraints that are required for type checking a Plutus AST.
+type MonadTypeCheck err term uni fun ann m =
```

Comment:
> Why is the `ReaderT` bit kept concrete but not the rest? I'm not sure I can quite put my finger on what the heuristics are here, but something feels off.

It's on purpose, I should've left a Note. We already have a similar one (Note [NormalizeTypeT]), but not for the type checker.

Basically, each procedure (renamer, type normalizer, type checker) has an `Internal` part and an API one. We use `transformers` for `internal \ external` and `mtl` for `external`  (unless we need generality even for the internal part, which is the case for the `renamer`, because we instantiate it to different monads in tests). This may seem weird, but there's a good reason: we can share the `external` constraints between the internal worker and the external wrapper. If a constraint is required by the API, then we make it `mtl`-style for generality (because the caller benefits from generality) and also reuse it in the internal part.

We could use `mtl` in both the cases, but:

1. I'm not a fan of using `mtl` unless there's a specific reason for that, because monads don't commute and `mtl` kinda pretends that the order doesn't matter
2. with the current way it's very clear what the internal and external parts are, it adds structure to otherwise a random pile of constraints

Reply:
I think maybe this should instead go in the style guide, since it reflects a general code structuring principle that we should probably try to follow generally?

Reply:
Good idea, created [PLT-400](https://input-output.atlassian.net/browse/PLT-400) for that and not gonna address it in this PR.